unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/BuildModeHUD/Scripts/BuildModeHUDController.cs

```diff
@@ -0,0 +1,380 @@
+using DCL.Controllers;
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class BuildModeHUDController : IHUD
+{
+    public event Action OnChangeModeAction;
+    public event Action OnTranslateSelectedAction;
+    public event Action OnRotateSelectedAction;
+    public event Action OnScaleSelectedAction;
+    public event Action OnResetAction;
+    public event Action OnDuplicateSelectedAction;
+    public event Action OnDeleteSelectedAction;
+    public event Action OnEntityListVisible;
+    public event Action OnStopInput;
+    public event Action OnResumeInput;
+    public event Action OnTutorialAction;
+    public event Action OnPublishAction;
+    public event Action OnLogoutAction;
+    public event Action<CatalogItem> OnCatalogItemSelected;
+    public event Action<DCLBuilderInWorldEntity> OnEntityClick;
+    public event Action<DCLBuilderInWorldEntity> OnEntityDelete;
+    public event Action<DCLBuilderInWorldEntity> OnEntityLock;
+    public event Action<DCLBuilderInWorldEntity> OnEntityChangeVisibility;
+    public event Action<DCLBuilderInWorldEntity, string> OnEntityRename;
+    public event Action<DCLBuilderInWorldEntity> OnEntitySmartItemComponentUpdate;
+    public event Action<Vector3> OnSelectedObjectPositionChange;
+    public event Action<Vector3> OnSelectedObjectRotationChange;
+    public event Action<Vector3> OnSelectedObjectScaleChange;
+    public event Action OnCatalogOpen; // Note(Adrian): This is used right now for tutorial purposes
+
+    internal IBuildModeHUDView view;
+
+    internal bool areExtraButtonsVisible = false,
+                  isControlsVisible = false, 
+                  isEntityListVisible = false, 
+                  isSceneLimitInfoVisibile = false,
+                  isCatalogOpen = false;
+
+    internal BuildModeHUDInitializationModel controllers;
+    internal CatalogItemDropController catalogItemDropController;
+
+    public BuildModeHUDController()
+    {
+        CreateBuildModeControllers();
+        CreateMainView(BuildModeHUDView.Create());
```

Comment:
I would refactor this a little bit. First I would move everything from the constructor to an `Initialize` method, this will improve our tests performance since you can create the controller, mock everything and then `Initialize`.

Another concern I have is related to `CreateMainView(BuildModeHUDView.Create());` line. Even though we can always call `CreateMainView` to replace the view with a stub the "real" view will be created at least once. We can use an approach like the following to avoid that.

```
public BuildModeHUDController()
{
	CreateBuildModeControllers();
        CreateMainView();
	...
}

public void CreateMainView()
{
	IBuildModeHUDView newView = CreateView();
	...
}

public virtual IBuildModeHUDView CreateView()
{
	//Whatever
}
```

Once all these changes have been made we will be able to do something like this in a test:
```
[Test]
public void TestSomething()
{
        hudView = Substitute.For<IBuildModeHUDView>();
	buildModeHUDController = Substitute.ForPartsOf<BuildModeHUDController>();
        buildModeHUDController.Configure().CreateView().Returns(info => hudView);
        buildModeHUDController.Initialize();
}
```

Reply:
After our conversation, I have applied your suggested changes here: https://github.com/decentraland/explorer/pull/2103/commits/b6b7193b0ad14e99f34fd1f71a454f49b024414a Thanks Alex!